cmd/envtool/tests.go

```diff
@@ -17,24 +17,187 @@ package main
 import (
 	"bufio"
 	"bytes"
+	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"sort"
 	"strings"
+	"time"
 
 	"go.uber.org/zap"
 	"golang.org/x/exp/maps"
 
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 )
 
+type action string
+
+//nolint:godot // false positive for unexported identifiers
+const (
+	// actionRun means the test has started running.
+	actionRun action = "run"
+	// actionPause means the test has been paused.
+	actionPause action = "pause"
+	// actionCont means the test has continued running.
+	actionCont action = "cont"
+	// actionPass means the test passed.
+	actionPass action = "pass"
+	// actionBench means the benchmark printed log output but did not fail.
+	actionBench action = "bench"
+	// actionFail means the test or benchmark failed.
+	actionFail action = "fail"
+	// actionOutput means the test printed output.
+	actionOutput action = "output"
+	// actionSkip means the test was skipped or the package contained no tests.
+	actionSkip action = "skip"
+)
+
+// Status represents the status of a single test.
+type Status string
+
+// Constants representing different test statuses.
+const (
+	Fail    Status = "fail"
+	Skip    Status = "skip"
+	Pass    Status = "pass"
+	Ignore  Status = "ignore"  // for fluky tests
+	Unknown Status = "unknown" // result can't be parsed
+)
+
+// TestResults represents the collection of results from multiple tests.
+type TestResults struct {
+	// Test results by full test name.
+	TestResults map[string]TestResult
+}
+
+// TestResult represents the outcome of a single test.
+type TestResult struct {
+	Status Status
+	Output string
+}
+
+type testEvent struct {
+	Time           time.Time `json:"Time"`
+	Action         action    `json:"Action"`
+	Package        string    `json:"Package"`
+	Test           string    `json:"Test"`
+	Output         string    `json:"Output"`
+	ElapsedSeconds float64   `json:"Elapsed"`
+}
+
+func (te testEvent) Elapsed() time.Duration {
+	return time.Duration(te.ElapsedSeconds * float64(time.Second))
+}
+
+func extractRootTestName(fullTestName string) string {
+	parts := strings.Split(fullTestName, "/")
+	return parts[0]
+}
+
+// execTestCommand runs test with given arguments.
+func execTestCommand(command string, args []string, totalTest int, logger *zap.SugaredLogger) error {
+	bin, err := exec.LookPath(command)
+	if err != nil {
+		return err
+	}
+	cmd := exec.Command(bin, args...)
+	logger.Debugf("Running %s", strings.Join(cmd.Args, " "))
+
+	cmd.Stderr = os.Stderr
+	p, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+
+	var testCounter int = 1
+	tested := make(map[string]bool)
+	var r io.Reader = p
+	d := json.NewDecoder(r)
```

Comment:
```suggestion
	d := json.NewDecoder(p)
```

`p` already implement `io.Reader`